prov/gni/src/gnix_ep.c

```diff
@@ -1084,6 +1084,142 @@ int gnix_ep_open(struct fid_domain *domain, struct fi_info *info,
 
 }
 
+static int __match_context(struct slist_entry *item, const void *arg)
+{
+	struct gnix_fab_req *req;
+
+	req = container_of(item, struct gnix_fab_req, slist);
+
+	GNIX_INFO(FI_LOG_EP_CTRL, "testing: req->context=%p context=%p item=%p\n",
+			req->user_context, arg, item);
+
+	return req->user_context == arg;
+}
+
+static inline struct gnix_fab_req *__find_tx_req(
+		struct gnix_fid_ep *ep,
+		void *context)
+{
+	struct gnix_fab_req *req = NULL;
+	struct slist_entry *entry;
+	struct gnix_vc *vc;
+
+	fastlock_acquire(&ep->vc_list_lock);
+	dlist_for_each(&ep->wc_vc_list, vc, entry)
+	{
+		GNIX_INFO(FI_LOG_EP_CTRL, "checking vc=%p\n", vc);
+		fastlock_acquire(&vc->tx_queue_lock);
+		GNIX_INFO(FI_LOG_EP_CTRL, "locked\n");
+		entry = slist_remove_first_match(&vc->tx_queue,
+				__match_context, context);
+		fastlock_release(&vc->tx_queue_lock);
+
+		if (entry) {
+			req = container_of(entry, struct gnix_fab_req, slist);
+			break;
+		}
+	}
+	fastlock_release(&ep->vc_list_lock);
+
```

Comment:
The lock could be removed earlier, but then the release of the lock would be conditional in two cases, rather than unconditional in one. I think the difference would be minimal but consider the possibility that the instruction pipeline of the CPU is going to be interrupted twice instead of once. 

Regardless, I can fix this if you prefer the other style.